synapse/api/auth.py

```diff
@@ -174,96 +174,60 @@ async def _wrapped_get_user_by_req(
 
             access_token = self.get_access_token_from_request(request)
 
-            (
-                user_id,
-                device_id,
-                app_service,
-            ) = await self._get_appservice_user_id_and_device_id(request)
-            if user_id and app_service:
-                if ip_addr and self._track_appservice_user_ips:
-                    await self.store.insert_client_ip(
-                        user_id=user_id,
-                        access_token=access_token,
-                        ip=ip_addr,
-                        user_agent=user_agent,
-                        device_id="dummy-device"
-                        if device_id is None
-                        else device_id,  # stubbed
-                    )
-
-                requester = create_requester(
-                    user_id, app_service=app_service, device_id=device_id
+            # First check if it could be a request from an appservice
+            requester = await self._get_appservice_user(request)
+            if not requester:
```

Comment:
Why are we no longer returning early here? Shouldn't we being doing exactly the same logic, just on `requester`?

Reply:
I did this because we were recording the IP in both cases, so I wanted to deduplicate that, and there is not much that we're doing for regular users that we don't do for app service ones

Reply:
Ah, OK. Is there any chance that we can split the refactor and the change into separate commits? Or is that just messy?

Reply:
It should be doable, if you don't mind a `git rebase` and potentially confusing GitHub

Reply:
I don't mind a rebase! I've just stared at that code for a bit now and my eyes keep glazing over

Reply:
Done! I think the whole thing is more or less reviewable commit-by-commit